crates/rattler/src/install/installer/mod.rs

```diff
@@ -0,0 +1,556 @@
+                        }
+                        Ok((cache_path, record))
+                    })
+                    .map_err(JoinError::try_into_panic)
+                    .map(|res| match res {
+                        Ok(Ok(result)) => Ok(Some(result)),
+                        Ok(Err(e)) => Err(e),
+                        Err(Ok(payload)) => std::panic::resume_unwind(payload),
+                        Err(Err(_err)) => Err(InstallerError::Cancelled),
+                    })
+                    .left_future()
+                } else {
+                    ready(Ok(None)).right_future()
+                };
+
+                // Uninstall the package if it was removed.
+                if let Some(record) = operation.record_to_remove() {
+                    let reporter = reporter
+                        .as_deref()
+                        .map(move |r| (r, r.on_unlink_start(idx, record)));
+                    unlink_package(prefix.as_ref(), record).await.map_err(|e| {
+                        InstallerError::UnlinkError(record.repodata_record.file_name.clone(), e)
+                    })?;
+                    if let Some((reporter, index)) = reporter {
+                        reporter.on_unlink_complete(index);
+                    }
+                }
+
+                // Install the package if it was fetched.
+                if let Some((cached_path, record)) = package_to_install.await? {
+                    let reporter = reporter
+                        .as_deref()
+                        .map(|r| (r, r.on_link_start(idx, &record)));
+                    link_package(
+                        &record,
+                        prefix.as_ref(),
+                        &cached_path,
+                        base_install_options.clone(),
+                        driver,
+                    )
+                    .await?;
+                    if let Some((reporter, index)) = reporter {
+                        reporter.on_link_complete(index);
+                    }
+                }
+
+                if let Some(reporter) = &reporter {
+                    reporter.on_transaction_operation_complete(idx);
+                }
+
+                Ok::<_, InstallerError>(())
+            };
+
+            pending_futures.push(operation_future);
+        }
+
+        // Wait for all transaction operations to finish
+        while let Some(result) = pending_futures.next().await {
+            result?;
+        }
+        drop(pending_futures);
+
+        // Post process the transaction
+        let post_process_result = driver
+            .post_process(&transaction, prefix.as_ref())
+            .map_err(InstallerError::PostProcessingFailed)?;
+
+        if let Some(reporter) = &self.reporter {
+            reporter.on_transaction_complete();
+        }
+
+        Ok(InstallationResult {
+            transaction,
+            pre_process_result,
+            post_process_result,
+        })
+    }
+}
+
+async fn link_package(
```

Comment:
should we make this (and the next function) pub? Could have some use cases.

Reply:
We could also add this behavior to the `link_package` function that is called inside this function. Maybe that makes more sense since its already public. WDYT?